src/sources/util/tcp.rs

```diff
@@ -219,28 +221,28 @@ async fn handle_stream<T>(
     }
 
     let mut _token = None;
-    let mut shutdown = Some(shutdown);
+    let mut shutdown_signal = Some(shutdown_signal);
     let mut reader = FramedRead::new(socket, source.decoder());
     stream::poll_fn(move |cx| {
-        if let Some(fut) = shutdown.as_mut() {
+        if let Some(fut) = shutdown_signal.as_mut() {
             match fut.poll_unpin(cx) {
                 Poll::Ready(token) => {
                     debug!("Start graceful shutdown.");
+
+                    let socket = reader.get_mut();
+
                     // Close our write part of TCP socket to signal the other side
                     // that it should stop writing and close the channel.
-                    let socket: Option<&TcpStream> = reader.get_ref().get_ref();
-                    if let Some(socket) = socket {
-                        if let Err(error) = socket.shutdown(std::net::Shutdown::Write) {
-                            warn!(message = "Failed in signalling to the other side to close the TCP channel.", %error);
+                    match socket.shutdown().now_or_never() {
+                        None => error!(message = "Failed shutting down TCP socket immediately."),
```

Comment:
This should be correct, as [`poll_shutdown`](https://docs.rs/tokio/1.3.0/src/tokio/net/tcp/stream.rs.html#1092-1095) on `TcpStream` shutdowns synchronously and returns `Poll::Ready(Ok(()))`.

Reply:
If that implementation changes will we catch it with a test? Seems unlikely that it'll be a problem, but worth considering.

Reply:
I went down that rabbit hole entirely...

Turns out the `poll_shutdown` implementation has been broken for `MaybeTlsIncomingStream`. Fixing that made the whole shutdown procedure more straight-forward.

Fixed with f870496.

Reply:
One weird thing I noticed while debugging: In the `tcp_shutdown_infinite_stream` test, `sources::util::tcp::handle_stream` has been entered twice. The first time it shutdown gracefully immediately, without reading (or receiving?) a single frame. That shutdown has not been triggered by the shutdown signal.

Might be worth investigating more deeply.

Reply:
False alarm, I figured out what was causing this: We have a `wait_for_tcp` utility in the test setup that briefly connects to the source 😄